### PR TITLE
Homework20: Vladislav Savchuk

### PR DESCRIPTION
**Pull request by VladislavSavchuk**
# Developed program tests from homework #11

- Bank deposit
- Library

1. Created a separate module for tests.
2. Imported functions from the module.
3. Wrote tests to check the application performance.
4. The tests check both positive and negative results.
5. The tests are run using _unittest_

### CODE REVIEW

File: homework/homework11/library.py [hunk 1]
```diff
@@ -20,18 +20,18 @@ def __str__(self):
...
-        return f'Book Description: {descr}'
+        return descr
 
     def reservation(self, user):
         """This method allows the user to reserve a book if it is not busy"""
         if not self.busy and not self.reserved:
             self.reserved = True
             self.reserved_by = user
-            return "Book has been successfully reserved"
+            return "reserved_success"
```
Comment (Stml89) [inline homework/homework11/library.py]:
```suggestion
            return True
```

File: homework/homework11/library.py [hunk 2]
```diff
@@ -41,39 +41,36 @@ def receiving(self, user):
             self.busy = True
             self.reserved = False
             self.reserved_by = None
-            return "The book is free. You can take the book"
+            return "taken_success"
```
Comment (Stml89) [inline homework/homework11/library.py]:
```suggestion
            return True
```

File: homework/homework11/bank_deposit.py [hunk 3]
```diff
@@ -31,12 +31,12 @@ def __init__(self, name: str):
...
-print(deposit_person2.calculate_monthly_capitalization())
+# print(person_1.name)
+# deposit_person1 = Bank(1000000, 120, 10)
+# print(deposit_person1)
+# print(deposit_person1.calculate_monthly_capitalization())
+# print()
+# print(person_2.name)
+# deposit_person2 = Bank(999999.99, 60, 10)
+# print(deposit_person2)
+# print(deposit_person2.calculate_monthly_capitalization())
```
Comment (Stml89) [inline homework/homework11/bank_deposit.py]: If you do not want to use something when module was imported - just use `if __name__ == __main__`. And start using logging module instead of `print`

File: homework/homework20/test_bank_deposit.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        logger.info("Finished a test case Bank")
+
+    def test_bank_initialization(self):
+        """Test initialization of the Bank class."""
+        logger.info("Testing Bank attributes")
+
+        bank = Bank(1500, 24, 3.5)
+        self.assertEqual(bank.deposit_amount, 1500)
+        self.assertEqual(bank.period, 24)
+        self.assertEqual(bank.rate, 3.5)
```
Comment (Stml89) [inline homework/homework20/test_bank_deposit.py]: add some log that says that test was successful, e.g.:
`logger.info("Successfully verified  <.......>")`

File: homework/homework20/test_bank_deposit.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+        # Test with zero period
+        bank = Bank(1000, 0, 5)
+        result = bank.calculate_monthly_capitalization()
+        self.assertEqual(result, 1000.0)
+
+        # Test with zero rate
+        bank = Bank(1000, 12, 0)
+        result = bank.calculate_monthly_capitalization()
+        self.assertEqual(result, 1000.0)
```
Comment (Stml89) [inline homework/homework20/test_bank_deposit.py]: 3 different test cases?

**Review by Stml89 — CHANGES_REQUESTED** [review]

**Comment by VladislavSavchuk** [comment]
fixed according to the comments @Stml89

File: homework/homework20/test_bank_deposit.py (hunk 5 above)
Comment (Stml89) [reply homework/homework20/test_bank_deposit.py]: Please divide this test into 3 TCs

File: homework/homework20/test_bank_deposit.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+        bank = Bank(1500, 24, 3.5)
+        assert bank.deposit_amount == 1500, \
+            f"Expected deposit_amount to be 1500, got {bank.deposit_amount}"
+        assert bank.period == 24, \
+            f"Expected period to be 24, got {bank.period}"
+        assert bank.rate == 3.5, \
+            f"Expected rate to be 3.5, got {bank.rate}"
+
+        logger.info("Successfully verified bank attributes")
```
Comment (Stml89) [inline homework/homework20/test_bank_deposit.py:39]: Nice

File: homework/homework20/test_bank_deposit.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        result = bank.calculate_monthly_capitalization()
+        self.assertEqual(result, 1608.6)
+
+        bank = Bank(5000, 36, 4)
+        result = bank.calculate_monthly_capitalization()
+        self.assertEqual(result, 5636.36)
+
+        bank = Bank(20000, 48, 5)
+        result = bank.calculate_monthly_capitalization()
+        self.assertEqual(result, 24417.91)
```
Comment (Stml89) [inline homework/homework20/test_bank_deposit.py]: what is the diff between those 3 tests?

File: homework/homework20/test_bank_deposit.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        """Test negative cases for monthly capitalization calculation."""
+        logger.info("Testing negative calculation of monthly capitalization""")
+
+        bank = Bank(1500, 24, 3.5)
+        result = bank.calculate_monthly_capitalization()
+        self.assertNotEqual(result, 1500)
+
+        bank = Bank(5000, 36, 4)
+        result = bank.calculate_monthly_capitalization()
+        self.assertNotEqual(result, 5000)
```
Comment (Stml89) [inline homework/homework20/test_bank_deposit.py]: are they testing the case thing?

File: homework/homework20/test_bank_deposit.py (hunk 6 above)
Comment (Stml89) [reply homework/homework20/test_bank_deposit.py:39]: add the same for other TCs as well

**Review by Stml89 — CHANGES_REQUESTED** [review]

**Comment by VladislavSavchuk** [comment]
@Stml89
- corrected the tests based on your comments.
- divided the tests of boundary values into 3 separate test cases
- added Successfully logger to each test
- removed 2 unnecessary tests in test_calculate_monthly_capitalization that tested the same thing (my mistake)
- updated negative cases
It seems to be looking better now, thank you.

**Review by Stml89 — APPROVED** [review]
